src/System.Private.ServiceModel/src/Internals/System/Runtime/TimeoutHelper.cs

```diff
@@ -31,38 +43,34 @@ public TimeoutHelper(TimeSpan timeout)
             _deadlineSet = (timeout == TimeSpan.MaxValue);
         }
 
-        // No locks as we expect this class to be used linearly. 
-        // If another CancellationTokenSource is created, we might have a CancellationToken outstanding
-        // that isn't cancelled if _cts.Cancel() is called. This happens only on the Abort paths, so it's not an issue. 
-        private void InitializeCancellationToken(TimeSpan timeout)
+        public async Task<CancellationToken> GetCancellationTokenAsync()
         {
-            if (timeout == TimeSpan.MaxValue || timeout == Timeout.InfiniteTimeSpan)
-            {
-                _cancellationToken = CancellationToken.None;
-            }
-            else if (timeout > TimeSpan.Zero)
+            if (!_cancellationTokenInitialized)
             {
-                _cts = new CancellationTokenSource();
-                _cancellationToken = _cts.Token;
-                TimeoutTokenSource.FromTimeout((int)timeout.TotalMilliseconds).Register(s_cancelOnTimeout, this);
-            }
-            else
-            {
-                _cancellationToken = new CancellationToken(true);
-            }
-            _cancellationTokenInitialized = true;
-        }
+                // No locks as we expect this class to be used linearly.
+                // If another CancellationTokenSource is created, we might have a CancellationToken outstanding
+                // that isn't cancelled if _cts.Cancel() is called. This happens only on the Abort paths, so it's not an issue.
 
-        public CancellationToken CancellationToken
-        {
-            get
-            {
-                if (!_cancellationTokenInitialized)
+                var timeout = RemainingTime();
+                if (timeout == TimeSpan.MaxValue || timeout == Timeout.InfiniteTimeSpan)
                 {
-                    InitializeCancellationToken(this.RemainingTime());
+                    _cancellationToken = CancellationToken.None;
                 }
-                return _cancellationToken;
+                else if (timeout > TimeSpan.Zero)
+                {
+                    _cts = new CancellationTokenSource();
```

Comment:
The commit mentioned this is just step 1 of 2, I was going to do this for the next step but I will just include it in this commit then.